### PR TITLE
Source tiktok marketing: migrate to api v1.3

### PR DESCRIPTION
**Pull request by davydov-d**
## What
Migrating to API version 1.3 since current version seems to have become unstable:
https://github.com/airbytehq/oncall/issues/402
https://github.com/airbytehq/airbyte/actions/runs/2881964037

## How
Adjusting data to existing stream schemas since we can not change schemas in GA

### CODE REVIEW

**Review by lazebnyi — APPROVED** [review]

File: airbyte-integrations/connectors/source-tiktok-marketing/source_tiktok_marketing/api_adapter.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+            container[field_name] = value if value is None else type_caster(value)
+
+    def adjust(self, record: Record):
+        self._adjust_deprecated_fields(record)
+        self._adjust_new_fields(record)
+        self._adjust_renamed_fields(record)
+        self._adjust_type_changes(record)
+
+
+class v13_to_v12_adapter(APIResponseAdapter):
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-tiktok-marketing/source_tiktok_marketing/api_adapter.py:98]: wtf tiktok...😅

File: airbyte-integrations/connectors/source-tiktok-marketing/source_tiktok_marketing/api_adapter.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+            ("included_pangle_audience_package_ids", "pangle_audience_package_include"),
+            ("excluded_pangle_audience_package_ids", "pangle_audience_package_exclude"),
+            ("catalog_authorized_bc_id", "catalog_authorized_bc"),
+            ("auto_targeting_enabled", "automated_targeting"),
+        ),
+        "report/integrated/get/": (("metrics/placement_type", "metrics/placement"),),
+    }
+    _type_changed_fields_by_stream = {
+        "ads": (
+            ("advertiser_id", int),
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-tiktok-marketing/source_tiktok_marketing/api_adapter.py:201]: is this change safe? do we know these can always be adjusted to int?

Comment (davydov-d) [reply airbyte-integrations/connectors/source-tiktok-marketing/source_tiktok_marketing/api_adapter.py:201]: No guarantees. TikTok says they changed the data type of IDs for the sake of compatibility, so I believe this is relatively safe

image

our options here are:
 - accept this change up until the incident :shrug: 
 - stay safe: do not adjust the data type but change the schema
 - go further - apply all the API changes to schema (then I'd close this PR and open a new one)

File: docs/integrations/sources/tiktok-marketing.md [hunk 3]
```diff
@@ -537,7 +537,8 @@ The connector is restricted by [requests limitation](https://ads.tiktok.com/mark
 
 | Version | Date       | Pull Request                                             | Subject                                                                                       |
 |:--------|:-----------|:---------------------------------------------------------|:----------------------------------------------------------------------------------------------|
-| 0.1.14  | 2022-06-29 | [13890](https://github.com/airbytehq/airbyte/pull/13890) | Removed granularity config option                                                             |                                               |
+| 0.1.15  | 2022-08-19 | [15786](https://github.com/airbytehq/airbyte/pull/15786) | Migrate to API version 1.3                                                                    |
```
Comment (sherifnada) [inline docs/integrations/sources/tiktok-marketing.md:540]:
```suggestion
| 0.1.15  | 2022-08-19 | [15786](https://github.com/airbytehq/airbyte/pull/15786) | Use API version 1.3                                                                    under the hood but transfomr it to match the schema of v1.2 in preparation for migrating to v1.3 |                                               
```

**Review by sherifnada — APPROVED** [review]
lgtm, I added a comment here https://github.com/airbytehq/oncall/issues/402 which emphasizes the need to bite the bullet and migrate to v1.3 relatively soon. 

If this change turns out to be unstable, we should just revert it and migrate to v1.3 completely (without transformation of the output schema), since this is actually a Major API bump (even if they don't call it that)

**Review by sherifnada — CHANGES_REQUESTED** [review]
@davydov-d apologies for the flip-flopping here, but after much discussion I think this is what we should do: 
* if I understand correctly, we believe migrating to v1.3 is going to improve the general stability and future proofing of the connector but is not necessarily going to fix any specific problem. Is this understanding correct? 
* In addition, migrating to v1.3 as done in this PR carries its own set of risks e.g: we may start failing irrecoverably on casting string columns to ints
* for those reasons, I think the right thing is to create a new Tiktok connector which uses the API v1.3, and deprecate the v1.2 connector within 30-45 days of releasing the new connector version.

The plan would be something like follows: 

1. Duplicate everything related to the existing Tiktok connector: codebase into a new directory source-tiktok-ads (renaming the previous one if needed), docs page, etc..
2. In the definitions YAML, rename the current tiktok to Tiktok Ads (Deprecated -- use the newer version) and add the new connector
3. Migrate the implementation of the new connector to use V1.3
4. Once we release v1.3, reach out to all cloud users telling them to switch all their connections to the new connector and they have 30-45 days before we delete the old one

what do you think?

**Comment by davydov-d** [comment]
> @davydov-d apologies for the flip-flopping here, but after much discussion I think this is what we should do:
> 
> * if I understand correctly, we believe migrating to v1.3 is going to improve the general stability and future proofing of the connector but is not necessarily going to fix any specific problem. Is this understanding correct?
> * In addition, migrating to v1.3 as done in this PR carries its own set of risks e.g: we may start failing irrecoverably on casting string columns to ints
> * for those reasons, I think the right thing is to create a new Tiktok connector which uses the API v1.3, and deprecate the v1.2 connector within 30-45 days of releasing the new connector version.
> 
> The plan would be something like follows:
> 
> 1. Duplicate everything related to the existing Tiktok connector: codebase into a new directory source-tiktok-ads (renaming the previous one if needed), docs page, etc..
> 2. In the definitions YAML, rename the current tiktok to Tiktok Ads (Deprecated -- use the newer version) and add the new connector
> 3. Migrate the implementation of the new connector to use V1.3
> 4. Once we release v1.3, reach out to all cloud users telling them to switch all their connections to the new connector and they have 30-45 days before we delete the old one
> 
> what do you think?

@sherifnada sounds good, no objections. I'm closing this PR

**Pull request closed without merging**